doc/dev/autograph.rst

```diff
@@ -0,0 +1,329 @@
+            def wire_loop(j):
+
+                @cond(x[j] > 0)
+                def trainable_gate():
+                    qml.RX(x[j], wires=j)
+
+                @trainable_gate.else_if(x[j] < 0)
+                def negative_gate():
+                    qml.RY(x[j], wires=j)
+
+                trainable_gate.otherwise(lambda: None)
+                trainable_gate()
+
+            def cnot_loop(j):
+                qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+
+            for_loop(0, 4, 1)(wire_loop)()
+            for_loop(0, 4, 1)(cnot_loop)()
+
+        for_loop(0, jnp.shape(weights)[0], 1)(layer_loop)()
+        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))
+
+>>> cost(weights, data)
+array(0.30455313)
+
+Currently, AutoGraph supports converting the following Python control
+flow statements:
+
+- ``if`` statements (including ``elif`` and ``else``)
+- ``for`` loops
+
+``while`` loops are currently not supported.
+
+Nested functions
+----------------
+
+AutoGraph will continue to work even when the qjit-compiled function
+itself calls nested functions. All functions called within the
+qjit-compiled function will also have Python control flow captured
+and converted by AutoGraph.
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            y = x ** 2
+        else:
+            y = x ** 3
+        return y
+
+    @qjit(autograph=True)
+    def g(x, n):
+        for i in range(n):
+            x = x + f(x)
+        return x
+
+>>> g(0.4, 6)
+array(22.14135448)
+
+One way to verify that the control flow is being correctly captured and
+converted is to examine the jaxpr representation of the compiled
+program:
+
+>>> g.jaxpr
+{ lambda ; a:f64[] b:i64[]. let
+    c:f64[] = qfor[
+      apply_reverse_transform=False
+      body_jaxpr={ lambda ; d:i64[] e:f64[]. let
+          f:bool[] = gt e 5.0
+          g:f64[] = qcond[
+            branch_jaxprs=[
+              { lambda ; a:f64[] b_:f64[]. let c:f64[] = integer_pow[y=2] a in (c,) },
+              { lambda ; a_:f64[] b:f64[]. let c:f64[] = integer_pow[y=3] b in (c,) }
+            ]
+          ] f e e
+          h:f64[] = add e g
+        in (h,) }
+      body_nconsts=0
+    ] 0 b 1 0 a
+  in (c,) }
+
+Here, we can see the for loop contained within the ``qcond`` operation, and
+the two branches of the ``if`` statement represented by the ``branch_jaxprs``
+list.
+
+If statements
+-------------
+
+While most ``if`` statements you may write in Python will be automatically
+converted, there are some important constraints and restrictions to be aware of.
+
+No 'early' returns
+~~~~~~~~~~~~~~~~~~
+
+The following pattern, where you return from an ``if`` statement early,
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            return x ** 2
+        return x ** 3
+
+will not be correctly captured by AutoGraph, and instead will be interpreted as
+
+.. code-block:: python
+
+    def f(x):
+        if x > 5:
+            x = x ** 2
+        return x ** 3
```

Comment:
@dime10 it seems that TF doesn't have this issue, so this might actually be a bug 🤔

Reply:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/control_flow.md#return-statements

Reply:
I might rewrite this section and simply say 'Return statements inside `if` statements are not yet supported. No error will occur, but the resulting function will not have the expected behaviour'.

Reply:
Yes that's right, it just hasn't been implemented yet. There are few other features that we don't have as well, like `break`